swarm.go

```diff
@@ -236,94 +246,180 @@ func (s *Swarm) Close() error {
-func (s *Swarm) ConnectionsToPeer(p peer.ID) []*Conn {
-	return wrapConns(s.swarm.ConnsWithGroup(p))
-}
+// ConnsToPeer returns all the live connections to peer.
+func (s *Swarm) ConnsToPeer(p peer.ID) []inet.Conn {
+	// TODO: Consider sorting the connection list best to worst. Currently,
+	// it's sorted oldest to newest.
+	s.conns.RLock()
+	defer s.conns.RUnlock()
+	conns := s.conns.m[p]
+	output := make([]inet.Conn, len(conns))
+	for i, c := range conns {
+		output[i] = c
+	}
+	return output
+}
+
+// bestConnToPeer returns the best connection to peer.
+func (s *Swarm) bestConnToPeer(p peer.ID) *Conn {
+	// Selects the best connection we have to the peer.
+	// TODO: Prefer some transports over others. Currently, we just select
+	// the newest non-closed connection with the most streams.
+	s.conns.RLock()
+	defer s.conns.RUnlock()
+
+	var best *Conn
+	bestLen := 0
+	for _, c := range s.conns.m[p] {
+		if c.conn.IsClosed() {
+			// We *will* garbage collect this soon anyways.
+			continue
+		}
+		c.streams.Lock()
+		cLen := len(c.streams.m)
+		c.streams.Unlock()
 
-func (s *Swarm) HaveConnsToPeer(p peer.ID) bool {
-	return len(s.swarm.ConnsWithGroup(p)) > 0
-}
+		if cLen >= bestLen {
+			best = c
+			bestLen = cLen
+		}
 
-// Connections returns a slice of all connections.
-func (s *Swarm) Connections() []*Conn {
-	return wrapConns(s.swarm.Conns())
+	}
+	return best
 }
 
-// CloseConnection removes a given peer from swarm + closes the connection
-func (s *Swarm) CloseConnection(p peer.ID) error {
-	conns := s.swarm.ConnsWithGroup(p) // boom.
-	for _, c := range conns {
-		c.Close()
+// Connectedness returns our "connectedness" state with the given peer.
+//
+// To check if we have an open connection, use `s.Connectedness(p) ==
+// inet.Connected`.
+func (s *Swarm) Connectedness(p peer.ID) inet.Connectedness {
+	if s.bestConnToPeer(p) != nil {
+		return inet.Connected
 	}
-	return nil
+	return inet.NotConnected
 }
 
-// Peers returns a copy of the set of peers swarm is connected to.
-func (s *Swarm) Peers() []peer.ID {
-	conns := s.Connections()
+// Conns returns a slice of all connections.
+func (s *Swarm) Conns() []inet.Conn {
+	s.conns.RLock()
+	defer s.conns.RUnlock()
 
-	seen := make(map[peer.ID]struct{}, len(conns))
-	peers := make([]peer.ID, 0, len(conns))
-	for _, c := range conns {
-		p := c.RemotePeer()
-		if _, found := seen[p]; found {
-			continue
+	conns := make([]inet.Conn, 0, len(s.conns.m))
+	for _, cs := range s.conns.m {
+		for _, c := range cs {
+			conns = append(conns, c)
+		}
+	}
+	return conns
+}
+
+// ClosePeer closes all connections to the given peer.
+func (s *Swarm) ClosePeer(p peer.ID) error {
+	conns := s.ConnsToPeer(p)
+	switch len(conns) {
+	case 0:
+		return nil
+	case 1:
+		return conns[0].Close()
+	default:
+		errCh := make(chan error)
+		for _, c := range conns {
+			go func(c inet.Conn) {
+				errCh <- c.Close()
+			}(c)
 		}
 
-		seen[p] = struct{}{}
+		var errs []string
```

Comment:
Is there a better way to do this? This kind of sucks.

Reply:
Try templates + a wait group.

Reply:
Yeah, wait group would be much nicer pattern here.